<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Gota Morota" />

<meta name="date" content="2018-05-23" />

<title>Deterministic formulas for genomic prediction</title>

<script src="lab3_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="lab3_files/bootstrap-3.3.6/css/bootstrap.min.css" rel="stylesheet" />
<script src="lab3_files/bootstrap-3.3.6/js/bootstrap.min.js"></script>
<script src="lab3_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<script src="lab3_files/navigation-1.1/tabsets.js"></script>
<script src="lab3_files/navigation-1.1/codefolding.js"></script>
<link href="lab3_files/magnific-popup-1.1.0/magnific-popup.css" rel="stylesheet" />
<script src="lab3_files/magnific-popup-1.1.0/jquery.magnific-popup.min.js"></script>
<link href="lab3_files/readthedown-0.1/readthedown.css" rel="stylesheet" />
<script src="lab3_files/readthedown-0.1/readthedown.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #ffffff; color: #1f1c1b; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; background-color: #ffffff; color: #a0a0a0; border-right: 1px solid #a0a0a0; }
td.sourceCode { padding-left: 5px; }
pre, code { color: #1f1c1b; background-color: #ffffff; }
code > span.kw { color: #1f1c1b; font-weight: bold; } /* Keyword */
code > span.dt { color: #0057ae; } /* DataType */
code > span.dv { color: #b08000; } /* DecVal */
code > span.bn { color: #b08000; } /* BaseN */
code > span.fl { color: #b08000; } /* Float */
code > span.cn { color: #aa5500; } /* Constant */
code > span.ch { color: #924c9d; } /* Char */
code > span.sc { color: #3daee9; } /* SpecialChar */
code > span.st { color: #bf0303; } /* String */
code > span.vs { color: #bf0303; } /* VerbatimString */
code > span.ss { color: #ff5500; } /* SpecialString */
code > span.im { color: #ff5500; } /* Import */
code > span.co { color: #898887; } /* Comment */
code > span.do { color: #607880; } /* Documentation */
code > span.an { color: #ca60ca; } /* Annotation */
code > span.cv { color: #0095ff; } /* CommentVar */
code > span.ot { color: #006e28; } /* Other */
code > span.fu { color: #644a9b; } /* Function */
code > span.va { color: #0057ae; } /* Variable */
code > span.cf { color: #1f1c1b; font-weight: bold; } /* ControlFlow */
code > span.op { color: #1f1c1b; } /* Operator */
code > span.bu { color: #644a9b; font-weight: bold; } /* BuiltIn */
code > span.ex { color: #0095ff; font-weight: bold; } /* Extension */
code > span.pp { color: #006e28; } /* Preprocessor */
code > span.at { color: #0057ae; } /* Attribute */
code > span.re { color: #0057ae; } /* RegionMarker */
code > span.in { color: #b08000; } /* Information */
code > span.wa { color: #bf0303; } /* Warning */
code > span.al { color: #bf0303; font-weight: bold; } /* Alert */
code > span.er { color: #bf0303; text-decoration: underline; } /* Error */
code > span. { color: #1f1c1b; } /* Normal */
</style>


</head>

<body>


<div id="content" data-toggle="wy-nav-shift">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<nav id="nav-top" role="navigation" aria-label="top navigation">
    <a role="button" href="#" data-toggle="wy-nav-top"><span class="glyphicon glyphicon-menu-hamburger"></span></a>
</nav>

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span class="text-muted">Code</span> <span class="text-muted caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All</a></li>
</ul>
</div>

<div id="header">
<h1 class="title">Deterministic formulas for genomic prediction</h1>
<h1 class="subtitle">Quantitative Genetics and Genomics Workshop <span class="citation">@ESALQ</span></h1>
</div>


<div id="table-of-contents">
    <h2><a href="#content">Deterministic formulas for genomic prediction</a></h2>
    <div id="text-table-of-contents">
      <ul>
      <li><a href="#population-parameter-based-deterministic-prediction">Population parameter-based deterministic prediction</a></li>
      <li><a href="#data">Data</a></li>
      <li><a href="#imputation">Imputation</a></li>
      <li><a href="#quality-control">Quality control</a></li>
      <li><a href="#centering-and-scaling">Centering and scaling</a></li>
      <li><a href="#g-matrix">G matrix</a></li>
      <li><a href="#relationship-based-deterministic-prediction-of-vanraden-2008">Relationship-based deterministic prediction of VanRaden (2008)</a></li>
      <li><a href="#leave-one-out-cross-validation">Leave-one-out cross-validation</a></li>
      </ul>
    </div>
</div>

<div id="main">
<div id="population-parameter-based-deterministic-prediction" class="section level2">
<h2>Population parameter-based deterministic prediction</h2>
<p>ShinyGPAS - Shiny Genomic Prediction Accuracy Simulator</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(shiny)
<span class="kw">library</span>(plotly)
<span class="kw">runApp</span>(<span class="st">&quot;app.R&quot;</span>)</code></pre></div>
</div>
<div id="data" class="section level2">
<h2>Data</h2>
<p>Resende Jr. et al. (2012) (<a href="http://dx.doi.org/10.1534/genetics.111.137026">DOI: 10.1534/genetics.111.137026</a>) analyzed 17 traits in loblolly pine (Pinus taeda) data, which include 951 individuals genotyped with 4853 SNPs. In this homework assignment, we will use the derregressed breeding values of crown width across the planting beds at age 6 (CWAC6). Download the <a href="http://www.genetics.org/highwire/filestream/412827/field_highwire_adjunct_files/7/Loblolly_Pine_Resende_.zip">zip</a> file and type the following code.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read phenotype and SNP files</span>
CWAC6 &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;Loblolly_Pine_Resende/Phenotypic_Data Folder/DATA_nassau_age6_CWAC.csv&quot;</span>, 
    <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
SNP &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;Loblolly_Pine_Resende/Snp_Data.csv&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)

<span class="co"># remove missing phenotypes</span>
na.index &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">is.na</span>(CWAC6<span class="op">$</span>Derregressed_BV))
CWAC6 &lt;-<span class="st"> </span>CWAC6[<span class="op">-</span>na.index, ]
SNP &lt;-<span class="st"> </span>SNP[<span class="op">-</span>na.index, ]
<span class="kw">table</span>(CWAC6<span class="op">$</span>Genotype <span class="op">==</span><span class="st"> </span>SNP[, <span class="dv">1</span>])

<span class="co"># phenotypes</span>
y &lt;-<span class="st"> </span>CWAC6<span class="op">$</span>Derregressed_BV
y &lt;-<span class="st"> </span><span class="kw">matrix</span>(y, <span class="dt">ncol =</span> <span class="dv">1</span>)

<span class="co"># markers</span>
SNP &lt;-<span class="st"> </span>SNP[, <span class="op">-</span><span class="dv">1</span>]  <span class="co"># 861 x 4853</span>
SNP[SNP <span class="op">==</span><span class="st"> </span><span class="op">-</span><span class="dv">9</span>] &lt;-<span class="st"> </span><span class="ot">NA</span></code></pre></div>
</div>
<div id="imputation" class="section level2">
<h2>Imputation</h2>
<p>Replace missing marker genotypes with mean values. Then store the marker genotypes in a matrix object <code>W</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">W &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">ncol =</span> <span class="kw">ncol</span>(SNP), <span class="dt">nrow =</span> <span class="kw">nrow</span>(SNP))
<span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(SNP)) {
    <span class="co"># cat(&#39;j = &#39;, j, &#39;\n&#39;)</span>
    W[, j] &lt;-<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">is.na</span>(SNP[, j]), <span class="kw">mean</span>(SNP[, j], <span class="dt">na.rm =</span> <span class="ot">TRUE</span>), SNP[, j])
}</code></pre></div>
</div>
<div id="quality-control" class="section level2">
<h2>Quality control</h2>
<p>Perform a quality control by removing markers with MAF &lt; 0.05. How many markers are removed? Save the filtered genotype matrix in <code>W2</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">freq &lt;-<span class="st"> </span><span class="kw">colMeans</span>(W)<span class="op">/</span><span class="dv">2</span>
maf &lt;-<span class="st"> </span><span class="kw">ifelse</span>(freq <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dv">1</span> <span class="op">-</span><span class="st"> </span>freq, freq)
maf.index &lt;-<span class="st"> </span><span class="kw">which</span>(maf <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)
<span class="kw">length</span>(maf.index)
W2 &lt;-<span class="st"> </span>W[, <span class="op">-</span>maf.index]</code></pre></div>
</div>
<div id="centering-and-scaling" class="section level2">
<h2>Centering and scaling</h2>
<p>Standardize the genotype matrix to have a mean of zero and variance of one. Save this matrix as <code>Xs</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Ws &lt;-<span class="st"> </span><span class="kw">scale</span>(W2, <span class="dt">center =</span> <span class="ot">TRUE</span>, <span class="dt">scale =</span> <span class="ot">TRUE</span>)

<span class="co"># dimensions</span>
n &lt;-<span class="st"> </span><span class="kw">nrow</span>(Ws)
m &lt;-<span class="st"> </span><span class="kw">ncol</span>(Ws)</code></pre></div>
</div>
<div id="g-matrix" class="section level2">
<h2>G matrix</h2>
<p>Compute the second genomic relationship matrix of VanRaden (2008) <code>G</code> using the entire markers. Then add a very small positive constant (e.g., 0.001) to the diagonal elements so that <code>G</code> matrix is invertible.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">G &lt;-<span class="st"> </span><span class="kw">tcrossprod</span>(Ws)<span class="op">/</span><span class="kw">ncol</span>(Ws)
G &lt;-<span class="st"> </span>G <span class="op">+</span><span class="st"> </span><span class="kw">diag</span>(n) <span class="op">*</span><span class="st"> </span><span class="fl">0.001</span></code></pre></div>
</div>
<div id="relationship-based-deterministic-prediction-of-vanraden-2008" class="section level2">
<h2>Relationship-based deterministic prediction of VanRaden (2008)</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># training-testing sets partitioning</span>
n.trn &lt;-<span class="st"> </span><span class="dv">761</span>
n.tst &lt;-<span class="st"> </span><span class="dv">100</span>
y.trn &lt;-<span class="st"> </span>y[<span class="dv">1</span><span class="op">:</span>n.trn]
y.tst &lt;-<span class="st"> </span>y[n.trn <span class="op">+</span><span class="st"> </span><span class="dv">1</span><span class="op">:</span>n.tst]
Ws.trn &lt;-<span class="st"> </span>Ws[<span class="dv">1</span><span class="op">:</span>n.trn, ]
Ws.tst &lt;-<span class="st"> </span>Ws[n.trn <span class="op">+</span><span class="st"> </span><span class="dv">1</span><span class="op">:</span>n.tst, ]

<span class="co"># G matrix</span>
Gtrn &lt;-<span class="st"> </span><span class="kw">tcrossprod</span>(Ws.trn)<span class="op">/</span><span class="kw">ncol</span>(Ws.trn)
Gtrn &lt;-<span class="st"> </span>Gtrn <span class="op">+</span><span class="st"> </span><span class="kw">diag</span>(n.trn) <span class="op">*</span><span class="st"> </span><span class="fl">0.001</span>
Gtst.trn &lt;-<span class="st"> </span><span class="kw">tcrossprod</span>(Ws.tst, Ws.trn)<span class="op">/</span><span class="kw">ncol</span>(Ws.tst)

<span class="co"># When training-testing sets relationship is equal to zero</span>
lambda &lt;-<span class="st"> </span><span class="fl">1.029561</span>  <span class="co"># fit.trn$Ve / fit.trn$Vu</span>
I &lt;-<span class="st"> </span><span class="kw">diag</span>(n.trn)
<span class="kw">sqrt</span>(<span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">ncol =</span> <span class="dv">761</span>, <span class="dt">nrow =</span> <span class="dv">1</span>) <span class="op">%*%</span><span class="st"> </span><span class="kw">solve</span>(Gtrn <span class="op">+</span><span class="st"> </span>I <span class="op">*</span><span class="st"> </span>lambda) <span class="op">%*%</span><span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, 
    <span class="dt">nrow =</span> <span class="dv">761</span>, <span class="dt">ncol =</span> <span class="dv">1</span>))

<span class="co"># fit GREML to obtain variance components</span>
<span class="kw">library</span>(rrBLUP)
fit.trn &lt;-<span class="st"> </span><span class="kw">mixed.solve</span>(<span class="dt">y =</span> y.trn, <span class="dt">K =</span> Gtrn)

<span class="co"># Compute deterministic prediction accuracy</span>
r &lt;-<span class="st"> </span><span class="kw">array</span>()
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n.tst) {
    <span class="kw">cat</span>(<span class="st">&quot;Running &quot;</span>, <span class="st">&quot;individual &quot;</span>, i, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)
    r[i] &lt;-<span class="st"> </span><span class="kw">sqrt</span>(<span class="kw">matrix</span>(Gtst.trn[i, ], <span class="dt">nrow =</span> <span class="dv">1</span>) <span class="op">%*%</span><span class="st"> </span><span class="kw">solve</span>(Gtrn <span class="op">+</span><span class="st"> </span>I <span class="op">*</span><span class="st"> </span>lambda) <span class="op">%*%</span><span class="st"> </span>
<span class="st">        </span><span class="kw">matrix</span>(Gtst.trn[i, ], <span class="dt">ncol =</span> <span class="dv">1</span>))
}

<span class="co"># Historgram of accuracy (r)</span>
<span class="kw">hist</span>(r, <span class="dt">breaks =</span> <span class="dv">25</span>, <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>))

<span class="co"># Average squared relationship vs. Reliability (r^2)</span>
<span class="kw">plot</span>(<span class="kw">apply</span>(Gtst.trn<span class="op">^</span><span class="dv">2</span>, <span class="dv">1</span>, median), r<span class="op">^</span><span class="dv">2</span>, <span class="dt">xlab =</span> <span class="st">&quot;Average squared relationship&quot;</span>, 
    <span class="dt">ylab =</span> <span class="st">&quot;Reliability&quot;</span>)
<span class="kw">abline</span>(<span class="kw">lm</span>(r<span class="op">^</span><span class="dv">2</span> <span class="op">~</span><span class="st"> </span><span class="kw">apply</span>(Gtst.trn<span class="op">^</span><span class="dv">2</span>, <span class="dv">1</span>, median)), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)</code></pre></div>
</div>
<div id="leave-one-out-cross-validation" class="section level2">
<h2>Leave-one-out cross-validation</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Do not run</span>
ghat &lt;-<span class="st"> </span><span class="kw">array</span>()
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n.tst) {
    yNA &lt;-<span class="st"> </span>y.trn
    yNA &lt;-<span class="st"> </span><span class="kw">c</span>(yNA, <span class="ot">NA</span>)
    <span class="kw">cat</span>(<span class="st">&quot;Running &quot;</span>, <span class="st">&quot;individual &quot;</span>, i, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)
    Gtrn2 &lt;-<span class="st"> </span><span class="kw">rbind</span>(Gtrn, Gtst.trn[i, ])
    Gtrn2 &lt;-<span class="st"> </span><span class="kw">cbind</span>(Gtrn2, <span class="kw">c</span>(Gtst.trn[i, ], G[i <span class="op">+</span><span class="st"> </span>n.trn, i <span class="op">+</span><span class="st"> </span>n.trn]))
    fit &lt;-<span class="st"> </span><span class="kw">mixed.solve</span>(yNA, <span class="dt">K =</span> Gtrn2)
    ghat[i] &lt;-<span class="st"> </span>fit<span class="op">$</span>u[n.trn <span class="op">+</span><span class="st"> </span><span class="dv">1</span>]
}

<span class="co"># square root of heritability</span>
h &lt;-<span class="st"> </span><span class="kw">sqrt</span>(fit<span class="op">$</span>Vu<span class="op">/</span>(fit<span class="op">$</span>Vu <span class="op">+</span><span class="st"> </span>fit<span class="op">$</span>Ve))

<span class="co"># genomic prediction accuracy - cor(g, ghat)/h</span>
<span class="kw">cor</span>(ghat, y.tst)<span class="op">/</span>h  <span class="co"># </span>

<span class="co"># comparison</span>
<span class="kw">mean</span>(r)</code></pre></div>
</div>
</div>


</div>

<div id="postamble" data-toggle="wy-nav-shift" class="status">
<p class="author"><span class="glyphicon glyphicon-user"></span> Gota Morota</p>
<p class="date"><span class="glyphicon glyphicon-calendar"></span> May 23, 2018</p>
</div>


<script>
$(document).ready(function () {
 	  });
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
